src/java.base/share/classes/com/sun/crypto/provider/TlsKeyMaterialGenerator.java

```diff
@@ -196,82 +201,97 @@ private SecretKey engineGenerateKey0() throws GeneralSecurityException {
         System.arraycopy(keyBlock, ofs, serverKeyBytes, 0, keyLength);
         ofs += keyLength;
 
-        if (isExportable == false) {
-            // cipher keys
-            clientCipherKey = new SecretKeySpec(clientKeyBytes, alg);
-            serverCipherKey = new SecretKeySpec(serverKeyBytes, alg);
-
-            // IV keys if needed.
-            if (ivLength != 0) {
-                byte[] tmp = new byte[ivLength];
-
-                System.arraycopy(keyBlock, ofs, tmp, 0, ivLength);
-                ofs += ivLength;
-                clientIv = new IvParameterSpec(tmp);
-
-                System.arraycopy(keyBlock, ofs, tmp, 0, ivLength);
-                ofs += ivLength;
-                serverIv = new IvParameterSpec(tmp);
-            }
-        } else {
-            // if exportable suites, calculate the alternate
-            // cipher key expansion and IV generation
-            if (protocolVersion >= 0x0302) {
-                // TLS 1.1+
-                throw new RuntimeException(
-                    "Internal Error:  TLS 1.1+ should not be negotiating" +
-                    "exportable ciphersuites");
-            } else if (protocolVersion == 0x0301) {
-                // TLS 1.0
-                byte[] seed = concat(clientRandom, serverRandom);
-
-                byte[] tmp = doTLS10PRF(clientKeyBytes,
-                    LABEL_CLIENT_WRITE_KEY, seed, expandedKeyLength, md5, sha);
-                clientCipherKey = new SecretKeySpec(tmp, alg);
-
-                tmp = doTLS10PRF(serverKeyBytes, LABEL_SERVER_WRITE_KEY, seed,
-                            expandedKeyLength, md5, sha);
-                serverCipherKey = new SecretKeySpec(tmp, alg);
+        try {
+            if (isExportable == false) {
+                // cipher keys
+                clientCipherKey = new SecretKeySpec(clientKeyBytes, alg);
+                serverCipherKey = new SecretKeySpec(serverKeyBytes, alg);
 
+                // IV keys if needed.
                 if (ivLength != 0) {
-                    tmp = new byte[ivLength];
-                    byte[] block = doTLS10PRF(null, LABEL_IV_BLOCK, seed,
-                                ivLength << 1, md5, sha);
-                    System.arraycopy(block, 0, tmp, 0, ivLength);
+                    byte[] tmp = new byte[ivLength];
+
+                    System.arraycopy(keyBlock, ofs, tmp, 0, ivLength);
+                    ofs += ivLength;
                     clientIv = new IvParameterSpec(tmp);
-                    System.arraycopy(block, ivLength, tmp, 0, ivLength);
+
+                    System.arraycopy(keyBlock, ofs, tmp, 0, ivLength);
+                    ofs += ivLength;
                     serverIv = new IvParameterSpec(tmp);
```

Comment:
Seems easier to just use the IvParameterSpec(byte[], int, int) constructor?

Reply:
Yes.